Archives/VolFile.cpp

```diff
@@ -389,103 +382,102 @@ namespace Archives
 
-		// Check next header tag ("volh" tag)
-		// ----------------------------------
-		// Note: the size of this section must be 0
-		if (ReadTag(8, "volh") != 0) {
-			return false;
-		}
-
-		// Check for fileName string table tag ("vols" tag)
-		// ------------------------------------------------
-		if ((m_StringTableLength = ReadTag(16, "vols")) == -1) {
-			return false;
-		}
-
-		m_ActualStringTableLength = *((int*)m_BaseOfFile + 6);	// Store actual length
-		m_StringTable = (char*)m_BaseOfFile + 28;				// Store pointer to table
+		ReadStringTable();
 
 		// Make sure the string table fits in the header (and next header fits too)
 		if (m_HeaderLength < m_StringTableLength + 20) {
 			return false;
 		}
 
-		temp = m_StringTableLength + 24;			// Calculate offset to next section
-		// Check for volume index tag ("voli" tag)
-		// ---------------------------------------
-		// Store the offset to the Index table while it's still calculated
-		m_Index = (IndexEntry*)((char*)m_BaseOfFile + temp + 8);
-		if ((m_IndexTableLength = ReadTag(temp, "voli")) == -1) {
-			return false;
+		m_IndexTableLength = ReadTag(std::array<char, 4> { 'v', 'o', 'l', 'i' });
+		if (m_IndexTableLength > 0) {
+			m_IndexEntries.resize(m_IndexTableLength / sizeof(IndexEntry));
+			archiveFileReader->Read(m_IndexEntries.data(), m_IndexTableLength);
 		}
 
 		// Make sure the index table fits in the header
 		if (m_HeaderLength < m_StringTableLength + m_IndexTableLength + 24) {
 			return false;
 		}
 
-		// Determine the number of index entries
-		// -------------------------------------
 		m_NumberOfIndexEntries = m_IndexTableLength / 14;
+		ReadPackedFileCount();
+
+		return true;
+	}
+
+	void VolFile::ReadStringTable()
+	{
+		uint32_t actualStringTableLength;
+		archiveFileReader->Read(&actualStringTableLength, sizeof(actualStringTableLength));
+
+		std::string charBuffer;
+		charBuffer.resize(actualStringTableLength);
+		archiveFileReader->Read(&charBuffer[0], actualStringTableLength);
+
+		m_StringTable.push_back("");
+		for (std::size_t i = 0; i < charBuffer.size(); i++)
+		{
+			if (charBuffer[i] == '\0') {
+				m_StringTable.push_back("");
+				continue;
+			}
+
+			m_StringTable[m_StringTable.size() - 1].push_back(charBuffer[i]);
+		}
+
+		m_StringTable.erase(m_StringTable.begin() + m_StringTable.size() - 1);
+	}
+
+	void VolFile::ReadPackedFileCount()
+	{
 		// Count the number of valid entries
-		for (temp = 0; temp < m_NumberOfIndexEntries; temp++)
+		uint32_t packedFileCount = 0;
+		for (packedFileCount; packedFileCount < m_NumberOfIndexEntries; packedFileCount++)
 		{
 			// Make sure entry is valid
-			if (m_Index[temp].fileNameOffset == UINT_MAX) {
+			if (m_IndexEntries[packedFileCount].fileNameOffset == UINT_MAX) {
 				break;
 			}
 		}
-		m_NumberOfPackedFiles = temp;		// Store the number of used index entries
-
-		return true;						// Header read successfully
+		m_NumberOfPackedFiles = packedFileCount;
```

Comment:
Consider using `m_NumberOfPackedFiles` directly in the loop and eliminating the local variable.

Alternatively, the loop could start with `for (; packedFileCount < m_numberOfIndexEntries; packedFileCount++)`.

Reply:
What do you think of eliminating m_NumberOfPackedFiles altogether and just using m_numberOfIndexEntries to represent both? I'm guessing they are separate values to support deleting a file from a volume without recreating the volume? However, in practice, I don't think this is ever done, so maybe not a valid use case?

Thanks,
Brett

Reply:
You may have a point here. We don't repack in-place.

The point of having that value separate is to track empty index entries. The empty entries could be used to add new files without needing to rewrite and shift the contents to make room for extra index entries. The string table also has a similar pair of values for used versus allocated size.

The other side of this is, you can delete files from a VOL and not need to repack the whole thing. Just mark the index as unused. Even removing the entry from the string table and the file contents would be optional, though probably desired. Bonus, if the packed file was at the end of the VOL, it would be easy to truncate the file in constant time, with no copying of data to fill in gaps to save space.

I guess the question is, do we want to ever support repacking in place? My thought is probably not, since it's kind of messy, and supporting it means having extra wasted allocated space.

Reply:
I think the answer is no repacking in place by either removing or adding a file. If I'm deleting a file, I want to reduce the disk size of the volume, which probably means repacking. If I'm adding a file, it should be added so as to support the binary search for the filename, which is probably easiest to support by repacking. 

The other reason I'm thinking of not supporting this in maintainability. This feature is fully untested and I'd rather not bother with trying out and supporting the different use cases.

For now, I'll just open an issue on this one so we can merge the branch.

-Brett

Reply:
Good call. I'm eager to see these new changes merged in.